File: sdks/py_sdk/sw4rm/clients/negotiation_room_store.py
"""Negotiation Room storage backend abstraction.

This module provides a pluggable storage layer for NegotiationRoomClient,
enabling shared state across multiple client instances.

The key insight is that NegotiationRoomClient previously used instance-local
storage (self._proposals, self._votes, self._decisions), which broke
multi-instance deployments. By extracting storage into a separate abstraction
that can be shared via dependency injection, we enable:

1. Multiple client instances sharing the same storage
2. Producer/Critic/Coordinator agents in different processes sharing state
3. Future gRPC integration where the server becomes the source of truth

For persistent storage backends (JSON file, Redis, PostgreSQL), see the
colony/stores module which provides optional persistence implementations.

Based on the PolicyStore pattern established in policy_store.py.
"""
from __future__ import annotations

import threading
from abc import ABC, abstractmethod
from typing import List, Optional, Protocol

from sw4rm.negotiation_types import (
    NegotiationProposal,
    NegotiationVote,
    NegotiationDecision,
)


class NegotiationRoomStore(Protocol):
    """Protocol defining the interface for negotiation room storage backends.

    All storage implementations must support these operations for proposals,
    votes, and decisions. Thread-safety requirements are implementation-specific.

    This follows the same pattern as PolicyStore from policy_store.py, enabling
    pluggable backends ranging from in-memory to file-based to gRPC-based.
    """

    def has_proposal(self, artifact_id: str) -> bool:
        """Check if a proposal exists for the given artifact_id.

        Args:
            artifact_id: Unique identifier for the artifact.

        Returns:
            True if a proposal exists, False otherwise.
        """
        ...

    def get_proposal(self, artifact_id: str) -> Optional[NegotiationProposal]:
        """Retrieve a proposal by artifact_id.

        Args:
            artifact_id: Unique identifier for the artifact.

        Returns:
            The proposal if found, None otherwise.
        """
        ...

    def save_proposal(self, proposal: NegotiationProposal) -> None:
        """Save a proposal to storage.

        Args:
            proposal: The proposal to save.

        Raises:
            ValueError: If a proposal with the same artifact_id already exists.
        """
        ...

    def list_proposals(
        self, negotiation_room_id: Optional[str] = None
    ) -> List[NegotiationProposal]:
        """List all proposals, optionally filtered by negotiation room.

        Args:
            negotiation_room_id: If provided, only return proposals for this room.

        Returns:
            List of proposals matching the filter criteria.
        """
        ...

    def get_votes(self, artifact_id: str) -> List[NegotiationVote]:
        """Retrieve all votes for a specific artifact.

        Args:
            artifact_id: The identifier of the artifact.

        Returns:
            List of all votes for the artifact (empty list if no votes).
        """
        ...

    def add_vote(self, vote: NegotiationVote) -> None:
        """Add a vote for an artifact.

        Args:
            vote: The vote to add.

        Raises:
            ValueError: If the critic has already voted for this artifact.
        """
        ...

    def get_decision(self, artifact_id: str) -> Optional[NegotiationDecision]:
        """Retrieve the decision for an artifact.

        Args:
            artifact_id: The identifier of the artifact.

        Returns:
            The decision if available, None otherwise.
        """
        ...

    def save_decision(self, decision: NegotiationDecision) -> None:
        """Save a decision for an artifact.

        Args:
            decision: The decision to save.

        Raises:
            ValueError: If a decision already exists for this artifact.
        """
        ...


class NegotiationRoomStoreABC(ABC):
    """Abstract base class for negotiation room storage backends.

    Provides the same interface as NegotiationRoomStore Protocol but as an ABC
    for implementations that prefer inheritance over structural typing.
    """

    @abstractmethod
    def has_proposal(self, artifact_id: str) -> bool:
        """Check if a proposal exists for the given artifact_id."""
        pass

    @abstractmethod
    def get_proposal(self, artifact_id: str) -> Optional[NegotiationProposal]:
        """Retrieve a proposal by artifact_id."""
        pass

    @abstractmethod
    def save_proposal(self, proposal: NegotiationProposal) -> None:
        """Save a proposal to storage."""
        pass

    @abstractmethod
    def list_proposals(
        self, negotiation_room_id: Optional[str] = None
    ) -> List[NegotiationProposal]:
        """List all proposals, optionally filtered by negotiation room."""
        pass

    @abstractmethod
    def get_votes(self, artifact_id: str) -> List[NegotiationVote]:
        """Retrieve all votes for a specific artifact."""
        pass

    @abstractmethod
    def add_vote(self, vote: NegotiationVote) -> None:
        """Add a vote for an artifact."""
        pass

    @abstractmethod
    def get_decision(self, artifact_id: str) -> Optional[NegotiationDecision]:
        """Retrieve the decision for an artifact."""
        pass

    @abstractmethod
    def save_decision(self, decision: NegotiationDecision) -> None:
        """Save a decision for an artifact."""
        pass


class InMemoryNegotiationRoomStore(NegotiationRoomStoreABC):
    """Thread-safe in-memory implementation of NegotiationRoomStore.

    Stores proposals, votes, and decisions in memory with thread-safe access.
    Data is lost when the process terminates. Suitable for testing and
    single-process deployments.

    Thread-safety: This implementation IS thread-safe. All operations are
    protected by a reentrant lock (RLock) to allow nested calls.

    Usage for shared state across multiple client instances:
        # Create a single shared store
        shared_store = InMemoryNegotiationRoomStore()

        # Pass to multiple clients
        producer_client = NegotiationRoomClient(store=shared_store)
        critic_client = NegotiationRoomClient(store=shared_store)
        coordinator_client = NegotiationRoomClient(store=shared_store)

        # Now all clients share the same state
    """

    def __init__(self) -> None:
        """Initialize an empty in-memory store with thread-safe access."""
        self._lock = threading.RLock()
        self._proposals: dict[str, NegotiationProposal] = {}
        self._votes: dict[str, list[NegotiationVote]] = {}
        self._decisions: dict[str, NegotiationDecision] = {}

    def has_proposal(self, artifact_id: str) -> bool:
        """Check if a proposal exists for the given artifact_id.

        Thread-safe.

        Args:
            artifact_id: Unique identifier for the artifact.

        Returns:
            True if a proposal exists, False otherwise.
        """
        with self._lock:
            return artifact_id in self._proposals

    def get_proposal(self, artifact_id: str) -> Optional[NegotiationProposal]:
        """Retrieve a proposal by artifact_id.

        Thread-safe.

        Args:
            artifact_id: Unique identifier for the artifact.

        Returns:
            The proposal if found, None otherwise.
        """
        with self._lock:
            return self._proposals.get(artifact_id)

    def save_proposal(self, proposal: NegotiationProposal) -> None:
        """Save a proposal to storage.

        Thread-safe.

        Args:
            proposal: The proposal to save.

        Raises:
            ValueError: If a proposal with the same artifact_id already exists.
        """
        with self._lock:
            artifact_id = proposal.artifact_id
            if artifact_id in self._proposals:
                raise ValueError(
                    f"Proposal with artifact_id '{artifact_id}' already exists"
                )
            self._proposals[artifact_id] = proposal
            self._votes[artifact_id] = []

    def list_proposals(
        self, negotiation_room_id: Optional[str] = None
    ) -> List[NegotiationProposal]:
        """List all proposals, optionally filtered by negotiation room.

        Thread-safe.

        Args:
            negotiation_room_id: If provided, only return proposals for this room.

        Returns:
            List of proposals matching the filter criteria.
        """
        with self._lock:
            if negotiation_room_id is None:
                return list(self._proposals.values())

            return [
                p
                for p in self._proposals.values()
                if p.negotiation_room_id == negotiation_room_id
            ]

    def get_votes(self, artifact_id: str) -> List[NegotiationVote]:
        """Retrieve all votes for a specific artifact.

        Thread-safe.

        Args:
            artifact_id: The identifier of the artifact.

        Returns:
            List of all votes for the artifact (empty list if no votes).
        """
        with self._lock:
            return list(self._votes.get(artifact_id, []))

    def add_vote(self, vote: NegotiationVote) -> None:
        """Add a vote for an artifact.

        Thread-safe.

        Args:
            vote: The vote to add.

        Raises:
            ValueError: If the critic has already voted for this artifact.
        """
        with self._lock:
            artifact_id = vote.artifact_id
            critic_id = vote.critic_id

            # Check if this critic has already voted
            existing_votes = self._votes.get(artifact_id, [])
            for existing_vote in existing_votes:
                if existing_vote.critic_id == critic_id:
                    raise ValueError(
                        f"Critic '{critic_id}' has already voted for "
                        f"artifact '{artifact_id}'"
                    )

            # Initialize votes list if needed (for cases where proposal
            # was loaded from external source)
            if artifact_id not in self._votes:
                self._votes[artifact_id] = []

            self._votes[artifact_id].append(vote)

    def get_decision(self, artifact_id: str) -> Optional[NegotiationDecision]:
        """Retrieve the decision for an artifact.

        Thread-safe.

        Args:
            artifact_id: The identifier of the artifact.

        Returns:
            The decision if available, None otherwise.
        """
        with self._lock:
            return self._decisions.get(artifact_id)

    def save_decision(self, decision: NegotiationDecision) -> None:
        """Save a decision for an artifact.

        Thread-safe.

        Args:
            decision: The decision to save.

        Raises:
            ValueError: If a decision already exists for this artifact.
        """
        with self._lock:
            artifact_id = decision.artifact_id
            if artifact_id in self._decisions:
                raise ValueError(
                    f"Decision already exists for artifact_id '{artifact_id}'"
                )
            self._decisions[artifact_id] = decision


# Singleton for easy sharing across the application
_default_store: Optional[InMemoryNegotiationRoomStore] = None
_default_store_lock = threading.Lock()


def get_default_store() -> InMemoryNegotiationRoomStore:
    """Get the default shared in-memory store.

    This provides a convenient way to share state across multiple
    NegotiationRoomClient instances within the same process without
    explicitly passing a store instance.

    Returns:
        The default shared InMemoryNegotiationRoomStore instance.

    Example:
        # All clients automatically share the same store
        client1 = NegotiationRoomClient()  # Uses default store
        client2 = NegotiationRoomClient()  # Uses same default store

        # Equivalent to:
        store = get_default_store()
        client1 = NegotiationRoomClient(store=store)
        client2 = NegotiationRoomClient(store=store)
    """
    global _default_store
    with _default_store_lock:
        if _default_store is None:
            _default_store = InMemoryNegotiationRoomStore()
        return _default_store


def reset_default_store() -> None:
    """Reset the default store (primarily for testing).

    This clears the default store singleton, causing a new store
    to be created on the next call to get_default_store().
    """
    global _default_store
    with _default_store_lock:
        _default_store = None

File: sdks/js_sdk/src/clients/negotiationRoomStore.ts
// Copyright 2025 Rahul Rajaram
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/**
 * Negotiation Room storage backend abstraction.
 *
 * This module provides a pluggable storage layer for NegotiationRoomClient,
 * enabling shared state across multiple client instances.
 *
 * The key insight is that NegotiationRoomClient previously used instance-local
 * storage (private Maps), which broke multi-instance deployments. By extracting
 * storage into a separate abstraction that can be shared, we enable:
 *
 * 1. Multiple client instances sharing the same storage
 * 2. Producer/Critic/Coordinator agents in different processes sharing state
 * 3. Future gRPC integration where the server becomes the source of truth
 *
 * For persistent storage backends (JSON file, Redis, PostgreSQL), see the
 * colony/stores module which provides optional persistence implementations.
 *
 * Based on the PolicyStore pattern established in policyStore.ts.
 */

import {
  NegotiationProposal,
  NegotiationVote,
  NegotiationDecision,
} from './negotiationRoom.js';

/**
 * Error thrown when a negotiation room store operation fails.
 */
export class NegotiationRoomStoreError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'NegotiationRoomStoreError';
  }
}

/**
 * Interface defining the contract for negotiation room storage backends.
 *
 * All storage implementations must support these operations for proposals,
 * votes, and decisions. This follows the same pattern as PolicyStore,
 * enabling pluggable backends ranging from in-memory to file-based to gRPC-based.
 */
export interface NegotiationRoomStore {
  /**
   * Check if a proposal exists for the given artifact_id.
   *
   * @param artifactId - Unique identifier for the artifact
   * @returns True if a proposal exists, false otherwise
   */
  hasProposal(artifactId: string): Promise<boolean>;

  /**
   * Retrieve a proposal by artifact_id.
   *
   * @param artifactId - Unique identifier for the artifact
   * @returns The proposal if found, null otherwise
   */
  getProposal(artifactId: string): Promise<NegotiationProposal | null>;

  /**
   * Save a proposal to storage.
   *
   * @param proposal - The proposal to save
   * @throws NegotiationRoomStoreError if a proposal with the same artifact_id already exists
   */
  saveProposal(proposal: NegotiationProposal): Promise<void>;

  /**
   * List all proposals, optionally filtered by negotiation room.
   *
   * @param negotiationRoomId - If provided, only return proposals for this room
   * @returns List of proposals matching the filter criteria
   */
  listProposals(negotiationRoomId?: string): Promise<NegotiationProposal[]>;

  /**
   * Retrieve all votes for a specific artifact.
   *
   * @param artifactId - The identifier of the artifact
   * @returns List of all votes for the artifact (empty array if no votes)
   */
  getVotes(artifactId: string): Promise<NegotiationVote[]>;

  /**
   * Add a vote for an artifact.
   *
   * @param vote - The vote to add
   * @throws NegotiationRoomStoreError if the critic has already voted for this artifact
   */
  addVote(vote: NegotiationVote): Promise<void>;

  /**
   * Retrieve the decision for an artifact.
   *
   * @param artifactId - The identifier of the artifact
   * @returns The decision if available, null otherwise
   */
  getDecision(artifactId: string): Promise<NegotiationDecision | null>;

  /**
   * Save a decision for an artifact.
   *
   * @param decision - The decision to save
   * @throws NegotiationRoomStoreError if a decision already exists for this artifact
   */
  saveDecision(decision: NegotiationDecision): Promise<void>;
}

/**
 * In-memory implementation of NegotiationRoomStore.
 *
 * Stores proposals, votes, and decisions in memory. Data is lost when the
 * process terminates. Suitable for testing and single-process deployments.
 *
 * Thread-safety: JavaScript is single-threaded, but async operations are
 * properly sequenced.
 *
 * @example
 * ```typescript
 * // Create a single shared store
 * const sharedStore = new InMemoryNegotiationRoomStore();
 *
 * // Pass to multiple clients
 * const producerClient = new NegotiationRoomClient({ store: sharedStore });
 * const criticClient = new NegotiationRoomClient({ store: sharedStore });
 * const coordinatorClient = new NegotiationRoomClient({ store: sharedStore });
 *
 * // Now all clients share the same state
 * ```
 */
export class InMemoryNegotiationRoomStore implements NegotiationRoomStore {
  private proposals: Map<string, NegotiationProposal> = new Map();
  private votes: Map<string, NegotiationVote[]> = new Map();
  private decisions: Map<string, NegotiationDecision> = new Map();

  async hasProposal(artifactId: string): Promise<boolean> {
    return this.proposals.has(artifactId);
  }

  async getProposal(artifactId: string): Promise<NegotiationProposal | null> {
    return this.proposals.get(artifactId) || null;
  }

  async saveProposal(proposal: NegotiationProposal): Promise<void> {
    const artifactId = proposal.artifactId;

    if (this.proposals.has(artifactId)) {
      throw new NegotiationRoomStoreError(
        `Proposal with artifact_id '${artifactId}' already exists`
      );
    }

    // Set createdAt if not provided
    const proposalWithTimestamp: NegotiationProposal = {
      ...proposal,
      createdAt: proposal.createdAt || new Date().toISOString(),
    };

    this.proposals.set(artifactId, proposalWithTimestamp);
    this.votes.set(artifactId, []);
  }

  async listProposals(negotiationRoomId?: string): Promise<NegotiationProposal[]> {
    const allProposals = Array.from(this.proposals.values());

    if (negotiationRoomId === undefined) {
      return allProposals;
    }

    return allProposals.filter((p) => p.negotiationRoomId === negotiationRoomId);
  }

  async getVotes(artifactId: string): Promise<NegotiationVote[]> {
    return [...(this.votes.get(artifactId) || [])];
  }

  async addVote(vote: NegotiationVote): Promise<void> {
    const artifactId = vote.artifactId;
    const criticId = vote.criticId;

    // Check if this critic has already voted
    const existingVotes = this.votes.get(artifactId) || [];
    for (const existingVote of existingVotes) {
      if (existingVote.criticId === criticId) {
        throw new NegotiationRoomStoreError(
          `Critic '${criticId}' has already voted for artifact '${artifactId}'`
        );
      }
    }

    // Set votedAt if not provided
    const voteWithTimestamp: NegotiationVote = {
      ...vote,
      votedAt: vote.votedAt || new Date().toISOString(),
    };

    // Initialize votes array if needed
    if (!this.votes.has(artifactId)) {
      this.votes.set(artifactId, []);
    }

    this.votes.get(artifactId)!.push(voteWithTimestamp);
  }

  async getDecision(artifactId: string): Promise<NegotiationDecision | null> {
    return this.decisions.get(artifactId) || null;
  }

  async saveDecision(decision: NegotiationDecision): Promise<void> {
    const artifactId = decision.artifactId;

    if (this.decisions.has(artifactId)) {
      throw new NegotiationRoomStoreError(
        `Decision already exists for artifact_id '${artifactId}'`
      );
    }

    // Set decidedAt if not provided
    const decisionWithTimestamp: NegotiationDecision = {
      ...decision,
      decidedAt: decision.decidedAt || new Date().toISOString(),
    };

    this.decisions.set(artifactId, decisionWithTimestamp);
  }
}

// Singleton for easy sharing across the application
let defaultStore: InMemoryNegotiationRoomStore | null = null;

/**
 * Get the default shared in-memory store.
 *
 * This provides a convenient way to share state across multiple
 * NegotiationRoomClient instances within the same process without
 * explicitly passing a store instance.
 *
 * @returns The default shared InMemoryNegotiationRoomStore instance
 *
 * @example
 * ```typescript
 * // All clients automatically share the same store
 * const client1 = new NegotiationRoomClient();  // Uses default store
 * const client2 = new NegotiationRoomClient();  // Uses same default store
 *
 * // Equivalent to:
 * const store = getDefaultStore();
 * const client1 = new NegotiationRoomClient({ store });
 * const client2 = new NegotiationRoomClient({ store });
 * ```
 */
export function getDefaultStore(): InMemoryNegotiationRoomStore {
  if (defaultStore === null) {
    defaultStore = new InMemoryNegotiationRoomStore();
  }
  return defaultStore;
}

/**
 * Reset the default store (primarily for testing).
 *
 * This clears the default store singleton, causing a new store
 * to be created on the next call to getDefaultStore().
 */
export function resetDefaultStore(): void {
  defaultStore = null;
}
